languages/python/atom/element.py

```
@@ -483,22 +559,144 @@ def wait_for_elements_healthy(self, element_list, retry_interval=HEALTHCHECK_RET
 
             time.sleep(retry_interval)
 
-    def command_loop(self):
-        """
-        Waits for command to be put in element's command stream.
-        Sends acknowledge to caller and then runs command.
-        Returns response with processed data to caller.
-        """
-        while not self._command_loop_shutdown.isSet():
+    def command_loop(self, n_procs=1, block=True, read_block_ms=1000):
+        """Main command execution event loop
+
+        For each worker process, performs the following event loop:
+            - Waits for command to be put in element's command stream consumer 
+              group
+            - Sends Acknowledge to caller and then runs command
+            - Returns Response with processed data to caller
+
+        Args:
+            n_procs (integer): Number of worker processes.  Each worker process 
+                               will pull work from the Element's shared command 
+                               consumer group (defaults to 1).
+            block (bool, optional): Wait for the response before returning 
+                                    from the function.
+                                       block.
+            read_block_ms (integer, optional): Number of milliseconds to block
+                                               for during a stream read insde of 
+                                               a command loop.
+        """
+        # update self._pid in case e.g. we were constructed in a parent thread but 
+        # `command_loop` was explicitly called as a sub-process
+        self._pid = os.getpid()
+        n_procs = int(n_procs)
+        assert n_procs > 0, "n_procs must be a positive integer"
+
+        self.processes = []
+        for i in range(n_procs):
```

Would it be worth adding a check for the number of currently active Threads, and then throwing a warning (or maybe even an error) if there's more than one active Thread?  My understanding is that this could help us identify odd bugs that come from `fork`ing a multi-threaded process (since the threads won't get copied).  Michael I think you already brought this up so maybe it's solved in some other way already, just wondering what you think about doing a check here.

---

Hey Cody,

Good call out.  I think we can easily add a debug level log line around if the [number of active threads](https://stackoverflow.com/questions/12476888/how-to-check-the-number-of-currently-running-threads-in-python) is greater than one.

IIRC the thing I was thinking about previously was whether or not a child thread would be able to perform a fork out when `command_loop` opens a subsequent process.

I have specifically tested a scenario similar to the above in unit testing where I open up a `Thread` that invokes `command_loop` and found it worked perfectly well.   That said, the test harness did not explicitly require thread state from the parent process to be copied over and that is very much still an issue.    However, in practice I think we could and arguably should easily avoid this situation for running `command_loop`s.

Agree that the debug line could be useful in the sorts of settings you're describing.

---

Cool, thanks, that all makes sense.  My feeling is that with your work we'll have enough avenues to implement whatever form of parallelism we need, I just want to have warnings or docs for any avenues that aren't guaranteed to work.  Sounds like we're on the same page :-)

---

There's definitely a balance to be struck within these changes between giving folks enough rope to get tangled up versus not being flexible enough :/.  We had similar trade-off discussions around using `os.fork` versus `multiprocessing.Process` and settled on the latter to err more on the latter such that we can fail fast.

I agree and think it's similar here, but possibly erring on the other side.  A quick call out to multiple active threads will only help alert ppl to possible issues while not hopefully getting in their way too much if they're trying something kind of fancy or outside well-tested paths.